File: sponge-jruby/examples/script/rb/correlators_builder.rb
# Sponge Knowledge base
# Correlator builders

java_import java.util.concurrent.atomic.AtomicInteger
java_import java.util.Collections
java_import java.util.LinkedHashMap

def onInit
    # Variables for assertions only
    $sponge.setVariable("hardwareFailureScriptCount", AtomicInteger.new(0))
    $sponge.setVariable("hardwareFailureScriptFinishCount", AtomicInteger.new(0))
    $sponge.setVariable("eventLogs", Collections.synchronizedMap(LinkedHashMap.new()))
end

def onLoad
    $sponge.enable(CorrelatorBuilder.new("SampleCorrelator").withEvents(["filesystemFailure", "diskFailure"]).withMaxInstances(1)
        .withOnAcceptAsFirst { |correlator, event| event.name == "filesystemFailure" }
        .withOnInit { |correlator| $sponge.getVariable("eventLogs").put(correlator.meta.name, []) }
        .withOnEvent { |correlator, event|
            eventLog = $sponge.getVariable("eventLogs").get(correlator.meta.name)

            eventLog  << event
            correlator.logger.debug("{} - event: {}, log: {}", correlator.hashCode(), event.name, eventLog.map(&:to_s))
            $sponge.getVariable("hardwareFailureScriptCount").incrementAndGet()
            if eventLog.length >= 4
                $sponge.getVariable("hardwareFailureScriptFinishCount").incrementAndGet()
                correlator.finish()
            end
        })
end

def onStartup
    $sponge.event("filesystemFailure").set("source", "server1").send()
    $sponge.event("diskFailure").set("source", "server1").send()
    $sponge.event("diskFailure").set("source", "server2").send()
    $sponge.event("diskFailure").set("source", "server1").send()
    $sponge.event("diskFailure").set("source", "server2").send()
end

File: sponge-kotlin/src/test/kotlin/examples/CorrelatorsBuilder.kt
/*
 * Sponge Knowledge Base
 * Correlator builders
 */

package org.openksavi.sponge.kotlin.examples

import org.openksavi.sponge.event.Event
import org.openksavi.sponge.kotlin.KAction
import org.openksavi.sponge.kotlin.KCorrelator
import org.openksavi.sponge.kotlin.KCorrelatorBuilder
import org.openksavi.sponge.kotlin.KKnowledgeBase
import java.util.concurrent.atomic.AtomicInteger
import java.util.Collections

class CorrelatorsBuilder : KKnowledgeBase() {

    override fun onInit() {
        // Variables for assertions only
        sponge.setVariable("hardwareFailureScriptCount", AtomicInteger(0))
        sponge.setVariable("hardwareFailureScriptFinishCount", AtomicInteger(0))
        sponge.setVariable("eventLogs", Collections.synchronizedMap(mutableMapOf<String, MutableList<Event>>()))
    }

    override fun onLoad() {
        sponge.enable(KCorrelatorBuilder("SampleCorrelator").withEvents("filesystemFailure", "diskFailure").withMaxInstances(1)
                .withOnAcceptAsFirst({ correlator, event -> event.name == "filesystemFailure"})
                .withOnInit({ correlator -> sponge.getVariable<MutableMap<String, MutableList<Event>>>("eventLogs")[correlator.meta.name] = mutableListOf<Event>() })
                .withOnEvent({ correlator, event ->
                    val eventLog = sponge.getVariable<MutableMap<String, MutableList<Event>>>("eventLogs").get(correlator.meta.name)

                    eventLog!!.add(event)
                    logger.debug("{} - event: {}, log: {}", hashCode(), event.name, eventLog)
                    sponge.getVariable(AtomicInteger::class, "hardwareFailureScriptCount").incrementAndGet()
                    if (eventLog.size == 4) {
                        sponge.getVariable(AtomicInteger::class, "hardwareFailureScriptFinishCount").incrementAndGet()
                        correlator.finish()
                    }
                }))
    }

    override fun onStartup() {
        with(sponge) {
            event("filesystemFailure").set("source", "server1").send()
            event("diskFailure").set("source", "server1").send()
            event("diskFailure").set("source", "server2").send()
            event("diskFailure").set("source", "server1").send()
            event("diskFailure").set("source", "server2").send()
        }
    }
}
